File: 665-non-decreasing-array/665-non-decreasing-array.py
class Solution:
    def checkPossibility(self, nums: List[int]) -> bool:
        if(len(nums)<=2):
            return True
        past1=nums[0]
        past2=-1e5-1;
        count=0
        for i in range(1,len(nums)):
            if(past1>nums[i]):
                if(i+1==len(nums) and count==0):
                    return True
                elif(i+1<len(nums) and past1<=nums[i+1]):
                    past2=past1
                    past1=nums[i+1]
                    count+=1
                elif(past2<=nums[i]):
                    count+=1
                    past1=nums[i]
                else:
                    return 0
            else:
                past2=past1
                past1=nums[i]
        return count<=1

File: 665-non-decreasing-array/665-non-decreasing-array.cpp
class Solution {
public:
    bool checkPossibility(vector<int>& nums) {
        if(nums.size()<=2)
        {
            return true;
        }        
        int past1=nums[0];
        int past2=INT_MIN;
        int count=0;
        for(int i=1;i<nums.size();i++)
        {
           if(past1>nums[i])
           {
               if(i+1==nums.size() and count==0)
               {
                   return true;
               }
               if(i+1<nums.size() and past1<=nums[i+1])
               {
                   // do nothing
                   count++;
                   int temp=past1;
                   past1=nums[i+1];
                   past2=temp;
               }
               else if(nums[i]>=past2)
               {
                   count++;
                   int temp=past1;
                   past1=nums[i];
               }
               else
               {
                   return false;
               }
           }
           else
           {
               int temp=past1;
               past1=nums[i];
               past2=temp;
           }
        }
        return count<=1;
    }
};